File: Week 6/pset6/vigenere.py
#!/usr/bin/env python3
# ----------------------------------------------------------
# |     University of Amsterdam: Minor Programmeren        |
# |     Problem Set 6: Vigenere (less)                     |
# |     Created by Davy Boekhout on 09/10/2017             |
# |                                                        |
# |     Using Vigenere cipher, takes a plaintext string    |
# |     and returns the ciphertext.                        |
# ----------------------------------------------------------

import sys

ALPHA_LENGTH = 26


def encrypt(text, keyword):
    """Take in plaintext and return ciphertext using vigenere cipher"""
    # list comprehension creates numerical version of keyword (where a/A = 0)
    num_key = [ord(keyword[x]) - ord('A') if keyword[x].isupper() else
               ord(keyword[x]) - ord('a') for x in range(len(keyword))]

    # variable that's going to contain the ciphertext
    result = ""

    # declare key index and use length of keyword to wrap around
    k_len = len(keyword)
    k_idx = 0

    # loop over plaintext till we are done
    for i in range(len(text)):
        # encrypt if i'th character in text is in the alphabet
        if text[i].isalpha():
            # preserve case in ciphertext
            if text[i].islower():
                # add text[i] + key[k_idx] to result, wrap around if needed
                if (ord(text[i]) + num_key[k_idx % k_len]) > ord('z'):
                    result += chr(ord(text[i]) + num_key[k_idx % k_len] - ALPHA_LENGTH)
                    k_idx += 1
                else:
                    result += chr(ord(text[i]) + num_key[k_idx % k_len])
                    k_idx += 1

            if text[i].isupper():
                if (ord(text[i]) + num_key[k_idx % k_len]) > ord('Z'):
                    result += chr(ord(text[i]) + num_key[k_idx % k_len] - ALPHA_LENGTH)
                    k_idx += 1
                else:
                    result += chr(ord(text[i]) + num_key[k_idx % k_len])
                    k_idx += 1
        # leave non alphabetical characters unchanged
        else:
            result += text[i]
    return result


def main():
    # validate input
    if len(sys.argv) == 2 and sys.argv[1].isalpha():
        # get plaintext from the user
        plaintext = str(input("plaintext: "))

        # encrypt plaintext
        crypttext = encrypt(plaintext, sys.argv[1])

        # print the result to stdout
        print("ciphertext: {}".format(crypttext))

    # exit if user does not comply
    else:
        print("Usage: vigenere.py <alphabetical_key>")
        exit(1)


if __name__ == '__main__':
    main()

File: Week 2/pset2/vigenere/vigenere.c
/*
----------------------------------------------------------
|     University of Amsterdam: Minor Programmeren        |
|     Problem Set 2: Vigenere                            |
|     Created by Davy Boekhout on 11/09/2017             |
|                                                        |
|     Using Vigenere cipher, takes a plaintext string    |
|     and returns the ciphertext.                        |
----------------------------------------------------------
*/

#define ALPHA_LENGTH 26

#include <stdio.h>
#include <ctype.h>
#include <string.h>
#include <stdlib.h>

void encrypt(char *plaintext, char *key);

int main(int argc, char * argv[])
{
    // validate user input
    if (argc == 2)
    {
        // reserve space for the key and loop over every character
        char key[strlen(argv[1])];
        for (int i = 0, n = strlen(argv[1]); i < n; i++)
        {
            // determine if the i'th character is valid
            if (!isalpha(argv[1][i]))
            {
                printf("ERROR: Keyword may only consist of alphanumeric characters\n");
                return 1;
            }
            // while input is valid, create the key string
            else
                key[i] = argv[1][i];
        }

        // key is valid so we can retrieve the plaintext
        char plaintext[128];
        printf("plaintext: ");
        fgets(plaintext, sizeof(plaintext), stdin);
        // after retrieving the plaintext, call encrypt() which prints the ciphertext
        if (plaintext[strlen(plaintext) - 1] == '\n')
            encrypt(plaintext,key);
    }
    // quit on mismatching argument count
    else
        return 1;
}

void encrypt(char *plaintext, char *key)
{
    // determine the numerical key shift for every character in key
    int length = strlen(key);
    for (int k = 0; k < length; k++)
    {
        if(isupper(key[k]))
            key[k] = key[k] - 'A';
        else
            key[k] = key[k] - 'a';
    }

    printf("ciphertext: ");

    // declare i for looping over the plaintext, j for looping over the key
    int i = 0;
    int j = 0;
    while (i < strlen(plaintext))
    {
        // determine if i'th character is a letter
        if (isalpha(plaintext[i]))
        {
            // determine if i'th character is upper or lower case
            if (islower(plaintext[i]))
            {
                // shifting the i'th character of the plaintext with the shift of key[j]
                // if plaintext[i] + key[j] is greater than 122, we need to wrap around the alphabet
                if ((plaintext[i] + (key[j % length]) > 'z'))
                {
                    printf("%c", (plaintext[i] + (key[j % length]) - ALPHA_LENGTH));
                    j++;
                }
                // if plaintext[i] + key[j] does not require wrapping around, continue
                else
                {
                    printf("%c", (plaintext[i] + (key[j % length])));
                    j++;
                }
            }
            // same rules apply as before, in this case for uppercase letters
            else if (isupper(plaintext[i]))
            {
                if ((plaintext[i] + (key[j % length]) > 'Z'))
                {
                    printf("%c", (plaintext[i] + (key[j % length]) - ALPHA_LENGTH));
                    j++;
                }
                else
                {
                    printf("%c", (plaintext[i] + (key[j % length])));
                    j++;
                }
            }
        }
        // print every character that does not match the criteria above
        else
            printf("%c", plaintext[i]);
        i++;
    }
}